### PR TITLE
go: implement smoke test

### PR DESCRIPTION
**Pull request by pongad**
Some shared code is modified so that request-object methods
work properly.

### CODE REVIEW

File: src/main/resources/com/google/api/codegen/go/smoke.snip [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  )
+
+  var _ = iterator.Done
+
+  func {@view.name}(t *testing.T) {
+    if testing.Short() {
+      t.Skip("skipping smoke test in short mode")
+    }
+
+    projectId := testutil.ProjID()
```
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/go/smoke.snip]: I have some concerns with the projectId here:
1. projectId is not always needed for every test
2. We should allow the user to pass in the ID from command line. I am not sure what it should look like in Go but in Java we use option args:
https://github.com/googleapis/toolkit/blob/master/src/test/java/com/google/api/codegen/testdata/java_smoke_test_library.baseline#L57

File: src/test/java/com/google/api/codegen/testdata/go_smoke_library.baseline [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      Name: formattedName,
+      Book: book,
+  }
+
+  resp, err := c.UpdateBook(ctx, request)
+  if err != nil {
+    t.Error(err)
+  } else {
+    t.Log(resp)
+  }
```
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/go_smoke_library.baseline]: Print "OK" if the test succeeded.

File: src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java [hunk 3]
```diff
@@ -226,9 +226,16 @@ private InitCodeContext createResponseInitCodeContext(
   }
 
   public TestCaseView createSmokeTestCaseView(GapicMethodContext context) {
-    ClientMethodType methodType = ClientMethodType.FlattenedMethod;
-    if (context.getMethodConfig().isPageStreaming()) {
+    GapicMethodConfig methodConfig = context.getMethodConfig();
+    ClientMethodType methodType = null;
+    if (methodConfig.isPageStreaming() && context.isFlattenedMethodContext()) {
```
Comment (pongad) [inline src/main/java/com/google/api/codegen/transformer/TestCaseTransformer.java]: Done

File: src/main/resources/com/google/api/codegen/go/smoke.snip (hunk 1 above)
Comment (pongad) [inline src/main/resources/com/google/api/codegen/go/smoke.snip]: For context: The `testutil` package is already used by handwritten clients. We're just reusing it.

> 2. We should allow the user to pass in the ID from command line

It is configurable with `GCLOUD_TESTS_GOLANG_PROJECT_ID` environment variable. So you can run
```shell
GCLOUD_TESTS_GOLANG_PROJECT_ID=my-project go test cloud.google.com/go/...
```
and test all handwritten and gapic clients with the same project.

> 1. projectId is not always needed for every test

It's OK. If you don't set the env var, `testutil.ProjID` returns empty string. If the test doesn't need it, it doesn't matter. It won't throw exception and fail the test.

File: src/test/java/com/google/api/codegen/testdata/go_smoke_library.baseline (hunk 2 above)
Comment (pongad) [inline src/test/java/com/google/api/codegen/testdata/go_smoke_library.baseline]: The `go test` tool already does it when tests succeed. For example:
```
ok  	cloud.google.com/go/speech/apiv1	0.133s
ok  	cloud.google.com/go/speech/apiv1beta1	0.120s
ok  	cloud.google.com/go/trace/apiv1	2.675s
ok  	cloud.google.com/go/vision/apiv1	3.230s
```

**Comment by pongad** [comment]
@shinfan PTAL

File: src/main/resources/com/google/api/codegen/go/smoke.snip (hunk 1 above)
Comment (shinfan) [inline src/main/resources/com/google/api/codegen/go/smoke.snip]: Ok fair enough

**Comment by pongad** [comment]
Sorry for the churn: as generated the smoke test will work on my desktop because it has application default cred. However, it won't work on CI, because creds are passed in another way. https://github.com/googleapis/toolkit/pull/1238 will help with this.

**Comment by pongad** [comment]
@shinfan @jba PTAL

**Comment by jba** [comment]
To avoid a lot of churn, you can remove the t.Log in the future.

**Comment by pongad** [comment]
@jba sounds good, I'll merge this now for expedience. Reported your concern at https://github.com/googleapis/toolkit/issues/1249 so I don't loose track.

**Pull request merged**